pep-9999.rst

```diff
@@ -0,0 +1,223 @@
+give more accurate reports.
+
+The motivation described in this PEP is an extension to :pep:`610`.  Besides
+stating information about packages installed using a direct URL, installers SHOULD
+record information also for packages installed from Python package indexes when
+identified by their name, and optionally their version.
+
+Specification
+=============
+
+The ``provenance_url.json`` file SHOULD be created in the ``*.dist-info``
+directory by installers when installing a distribution identified by their
+name, and optionally their version specifier.
+
+This file MUST NOT be created when installing a distribution from a requirement
+specifying a direct URL reference (including a VCS URL).
+
+Only one of ``provenance_url.json`` and ``direct_url.json`` from :pep:`610`
+files MAY be present in ``*.dist-info`` directory.
+
+The ``provenance_url.json`` JSON file MUST be a dictionary, compliant with
+:rfc:`8259` and UTF-8 encoded.
+
+If present, it MUST contain exactly two keys. The first one is ``url``, with
+type ``string``.  The second key MUST be ``archive_info`` with a value defined
+below.
+
+Following :pep:`610`, the ``url`` field MUST be stripped of any sensitive
+authentication information, for security reasons.
+
+The user:password section of the URL MAY however be composed of environment
+variables, matching the following regular expression::
+
+    \$\{[A-Za-z0-9-_]+\}(:\$\{[A-Za-z0-9-_]+\})?
+
+Additionally, the user:password section of the URL MAY be a well-known,
+non-security sensitive string. A typical example is ``git`` in the case of an
+URL such as ``ssh://git@gitlab.com``.
+
+The ``archive_info`` key MUST be present as a dictionary and MUST state at
+least one of the following keys:
+
+- The ``hashes`` key SHOULD be present as a dictionary mapping a hash name to a
+  hex-encoded digest of the file.
+
+  Multiple hashes can be included, and it is up to the consumer to decide what
+  to do with multiple hashes (it may validate all of them or a subset of them, or
+  nothing at all).
+
+  These hash names SHOULD always be normalized to be lowercase.
+
+  Any hash algorithm available via ``hashlib`` (specifically any that can be
+  passed to ``hashlib.new()`` and do not require additional parameters) can be
+  used as a key for the hashes dictionary. At least one secure algorithm from
+  ``hashlib.algorithms_guaranteed`` SHOULD always be included. At the time of
+  writing, ``sha256`` specifically is recommended.
+
+- A ``hash`` key (type ``string``) MAY be present for compatibility purposes
+  with :pep:`610`, with value ``<hash-algorithm>=<expected-hash>``.
+
+Producers of the data structure SHOULD emit the ``hashes``. Producers SHOULD
+continue to emit the ``hash`` key in contexts where they did so before, so as
+to keep compatibility with :pep:`610` for existing clients.
+
+When both the ``hash`` and ``hashes`` keys are present, the hash represented in
+the ``hash`` key MUST also be present in the ``hashes`` dictionary, so
```

Comment:
Very good!  This is important.  But this is the implementation complexity I'd rather we not even have.  Code can be written that deals with this JSON but does not do this and does not validate equality of hashes specified in more than one place, and blindly uses the simpler `hash` field... which is then a security problem.  get rid of the legacy field entirely in this file and the chance of that drops.

granted someone is still going to include the field.  and will pass these data structures elsewhere to code that uses that instead of a hashes dict.  but unless we require that code parsing these validate explicitly that the legacy `"hash"` key is not present, that unintended consequences bug cannot be avoided.

Equivalance validation logic is already complicated because hash names can be spelled in multiple ways.  How is validating the used name defined?  Do not trust `hashlib.new()` to do it for you.  It is not a canonical name validator.

Reply:
Sounds reasonable, incorporated in [3c9b67f](https://github.com/fridex/peps/pull/2/commits/3c9b67f50f1c3d56be53a5c89bd7cf1c8a2f836d).

Resolving this discussion, feel free to reopen if additional work is needed to address this. Thank you!